src/lib/libast/man/stk.3
.TH STK 3
.SH NAME
\fBstk\fR \- data stack storage library
.SH SYNOPSIS
.ta .75i 1.5i 2.25i 3i 3.75i 4.5i 5.25i 6i
.de Ss
.fl
.ne 3
.SS "\\$1"
..
.PP
.nf
\f3
#include <stk.h>

Stk_t *stkopen(int \fIflags\fP);
void stkoverflow(Stk_t *\fIstack\fP, void (*\fIoverflow\fP)(size_t));
Stk_t *stkinstall(Stk_t *\fIstack\fP, char *(*\fIoverflow\fP)(size_t));
int stkclose(Stk_t *\fIstack\fP);
unsigned int stklink(Stk_t *\fIstack\fP)

void *stkalloc(Stk_t *\fIstack\fP, size_t \fIsize\fP);
char *stkcopy(Stk_t *\fIstack\fP, const char *\fIstring\fP);
void *stkset(Stk_t *\fIstack\fP, void *\fIaddress\fP, ptrdiff_t \fIoffset\fP);

void *stkseek(Stk_t *\fIstack\fP, ptrdiff_t \fIoffset\fP);
ptrdiff_t stktell(Stk_t *\fIstack\fP);
char *stkptr(Stk_t *\fIstack\fP, ptrdiff_t \fIoffset\fP);
void *stkfreeze(Stk_t *\fIstack\fP, size_t \fIextra\fP);
\fR
.fi
.SH DESCRIPTION
.PP
\f3stk\fP is a package of routines designed to provide efficient
stack oriented dynamic storage.
A stack abstraction consists of an ordered list of contiguous
memory regions, called stack frames, that can hold objects of
arbitrary size.
A stack is represented by the type \f3Stk_t\fP
defined in the header \f3<stk.h>\fP.
The type \f3Stk_t\fP is compatible with the type \f3Sfio_t\fP
defined by the \f3sfio\fP(3) library.
There is a preset initial stack which can be referenced
by the constant \f3stkstd\fP (stack standard).
.PP
Objects of variable size can be
added to a stack,
and programs can reference these objects directly with pointers.
In addition, the last object on the stack
(referred to here as the current object)
can be built incrementally.
The current object has an associated offset that determines its
current size.
While the current object is being built incrementally,
its location might
change so that it is necessary to reference the object with
relative offsets ranging from zero to the current offset of the object.
.PP
The \f3sfio\fP(3) output functions can be used to build
the current object incrementally.
An object that is built incrementally on the stack will
always occupy contiguous memory within a stack frame,
but until \f3stkfreeze\fP() is called,
the location in memory for the object can change.
There is a current offset associated with the current object that
determines where subsequent operations apply.
Initially, this offset is zero, and the offset changes as a result
of the operations you specify.
.SS "Descriptions of the functions"
.PP
The \f3stkopen\fP() function creates a new stack.
The \fIflags\fP argument is an options bitmask.
If the \f3STK_SMALL\fP bit is set, the stack allocates memory in
small blocks, optimizing memory usage at the expense of performance.
If the \f3STK_NULL\fP bit is set, a stack overflow will cause stack
operations to return a null pointer instead of throwing an exception.
If the \f3STK_NULL\fP bit is not set and the \f3stkstd\fP stack has
a custom overflow function set (see \f3stkoverflow\fP() below),
then the new stack inherits the pointer to that overflow function.
If successful,
\f3stkopen\fP() returns a pointer to a stack whose reference
count is 1.
Otherwise, \f3stkopen\fP() returns a null pointer.
.PP
The \f3stkoverflow\fP() function specifies a function that will
be called whenever \f3malloc\fP(3) fails while trying to grow the
stack.
The \fIoverflow\fP function will be called with the size that was passed
to \f3malloc\fP(3).
The \fIoverflow\fP function can call \f3exit\fP(3), call \f3longjmp\fP(3)
or return.
If the \f2overflow\fP function returns,
normal failure behavior is resumed and the calling function will return
.BR NULL .
If \f2overflow\fP is \f3NULL\fP, then the stack overflow action
is reset to the default.
If the stack was opened with the \f3STK_NULL\fP option,
the default is to return a null pointer,
otherwise the default is to write an error to standard error and to
call \f3exit\fP(2) with a non-zero exit value.
.PP
The \f3stkinstall\fP() function
is deprecated and provided for backward compatibility.
\f3stkinstall\fP() makes \f3stkstd\fP refer to the specified \fIstack\fP pointer
and returns a pointer to the previous \f3stkstd\fP stack.
If the \fIoverflow\fP argument is not null and the stack was not opened with
the \f3STK_NULL\fP option, then \fIoverflow\fP sets the overflow action
(see \f3stkoverflow\fP() above, except that the function type is
.B "char*"
and not
.B void
for backward compatibility reasons;
the return pointer is no longer used).
When \fIstack\fP is a null pointer,
the \f3stkstd\fP stack is not changed
but the \fIoverflow\fP function for the \f3stkstd\fP stack can be changed
and a pointer to the \f3stkstd\fP stack is returned.
(To replace deprecated \f3stkinstall\fP() usage, simply
pass the pointer to the desired stack to the \f3stk\fP functions,
and use \f3stkoverflow\fP() to set the stack overflow function).
.PP
The \f3stkclose\fP() function decrements the reference count and
frees the memory associated with
the specified stack
when the reference count is zero.
The effect of subsequent references to objects
on the stack is undefined.
.PP
The \f3stklink\fP() function increases the reference count for the
given \fIstack\fP and returns the increased count.
.PP
The
\f3stkalloc\fP() function returns an aligned pointer to space on the
active stack that can be used to hold any object of the given \fIsize\fP.
\f3stkalloc\fP() is similar to \f3malloc\fP(3) except that individual
items returned by \f3stkalloc\fP() can not be freed.
Since 2026-07-29, \f3stkalloc\fP() initializes all allocated bytes to zero.
\f3stkalloc\fP() causes the offset of the current object to be set to
zero.
.PP
The \f3stkcopy\fP() function
copies the given \fIstring\fP onto the stack
as a new object,
terminates that object,
and returns a pointer to it.
If there is a current unterminated object,
it is relocated to the top of the stack,
with the new string object inserted below it.
.PP
The \f3stkset\fP() function finds the frame containing the given
\fIaddress\fP, frees all frames that were created after the one containing
the given \fIaddress\fP, and sets the current object to the given
\fIaddress\fP.
The top of the current object is set to \fIoffset\fP bytes from
current object.
If \fIaddress\fP is null, the stack is reset to the beginning.
If it is non-null, but is not the address of an object on the
stack, the program aborts and dumps core.
.PP
The \f3stkseek\fP() function is used set the offset for the
current object.
The \fIoffset\fP argument to \f3stkseek\fP() specifies the new
offset for the current object.
The frame will be extended or moved
if \f3offset\fP causes the new current offset to extend beyond the
current frame.
\f3stkseek\fP() returns a pointer to the beginning of the current object.
.PP
The \f3stktell\fP() function gives the current offset into the current
unterminated object.
.PP
The \f3stkptr\fP() function converts the given \f3offset\fP
for the current object into a memory address on the stack.
This address is only valid until another stack operation is given.
The result is not defined if \fIoffset\fP exceeds the size of the current
object.
This function returns a \f3char\fP pointer instead of \f3void\fP pointer
to make it easy to dereference the return value directly and read
an individual byte on the stack.
.PP
The \f3stkfreeze\fP()
function terminates the current object on the
stack and returns a pointer to the beginning of this object.
If \fIextra\fP is non-zero, \fIextra\fP bytes are added to the stack
before the current object is terminated.  The first added byte will
contain zero and the contents of the remaining bytes are undefined.
.SH HISTORY
The
\f3stk\fP
interface was derived from similar routines in the KornShell code
that is used for building parse trees and carrying out expansions.
It provides an efficient mechanism for grouping dynamically allocated
objects so that they can be freed all at once rather than individually.
.PP
In 2024, as part of changes made by
Martijn Dekker <martijn@inlv.org> for ksh 93u+m,
the return types of \f3stkalloc\fP(), \f3stkset\fP(), \f3stkseek\fP() and
\f3stkfreeze\fP() were changed from \f3char*\fP to \f3void*\fP,
the \f3stkoverflow\fP() function was added,
and the \f3stkinstall\fP() function was deprecated.
.SH AUTHOR
David Korn
.SH SEE ALSO
.IR exit (2),
.IR longjmp (3),
.IR malloc (3),
.IR sfio (3)


NEWS
This documents significant changes in the dev branch of ksh 93u+m.
For full details, see the git log at: https://github.com/ksh93/ksh
Uppercase BUG_* IDs are shell bug IDs as used by the Modernish shell library.

2026-09-18:

- Fixed: a globstar ('**') immediately following a symlink to a directory
  failed to be expanded, and thus failed to follow that symlink, if
  anything other than a bare '**' followed it in the pattern, e.g. in a
  pattern like 'symlink/**/file'. This is a follow-up fix to a previous,
  incomplete fix for this bug.

2026-09-17:

- Fixed multiple memory leaks occurring upon subshell exit when assigning a
  value to a variable with a shell discipline function within a subshell.

- Fixed a crash occurring when assigning, within a subshell, an invalid
  value to a variable of a type defined by the 'enum' command.

2026-09-16:

- Fixed: when the special IFS (field splitting characters) variable was
  assigned to within a ${ shared-state command substitution }, the state of IFS
  failed to persist beyond it, and a use after free (potential crash) occurred.

- Fixed bounds checking when assigning to fixed-size arrays (declared like,
  e.g., 'typeset -a array[4]'). Assigning to an out-of-range subscript now
  produces the expected error instead of potentially crashing.

2026-09-15:

- Fixed a memory leak when assigning, within a subshell, a value to a
  special variable (one of IFS, FPATH, CDPATH, SHELL, ENV, VISUAL, EDITOR,
  HISTFILE, OPTIND, SECONDS, RANDOM,.sh.match, .sh.version, LINENO, _,
  LC_CTYPE, LC_NUMERIC, LC_COLLATE, LC_MESSAGES, LC_ALL, and LANG).

2026-09-13:

- Fixed a memory leak that occurred when converting an existing scalar
  variable to an associative array using 'typeset -A'.

2026-09-12:

- Implemented advisory locking for $HISTFILE (~/.sh_history by default) so
  that multiple ksh instances can write to the history file without risking
  data loss when one of them decides to trim the file. Note that ksh
  versions before 2026-09-12 will ignore the lock, so the risk continues to
  exist if an older ksh binary shares write access to the file.

- History file trimming no longer requires write access to the directory in
  which the history file resides and will no longer change the directory's
  last-modified timestamp. However, it now requires a usable /tmp or $TMPDIR.

- I/O errors while writing to the history file are no longer silently ignored.

- Fixed a possible crash when attempting to set an integer variable to a
  multibyte character.

2026-09-10:

- Fixed a rare crash that could occur if a signal (such as interrupt, Ctrl+C)
  occurred while cleaning up the state of a virtual subshell upon its exit.

2026-09-08:

- Fixed: PWD and ENV were not properly reinitialised when executing a shell
  script without a #! path.

- Fixed a possible crash after unsetting RANDOM. Bug introduced on 2022-07-21.

2026-09-04:

- Fixed a longstanding bug in which output data loss or corruption could
  intermittently occur if, within a command substitution, standard error is
  redirected to standard output and, within that redirection, standard
  output is then connected to a pipe.

2026-09-02:

- Fixed a bug, introduced on 2025-05-30, that made 'print -s' ineffective
  at writing to the history file. This functionality is now restored.

2026-08-31:

- Fixed a crash in the expr(1) path-bound built-in command that could occur
  when using the non-standard 'match' or 'index' keyword.

2026-08-30:

- Fixed a bug that caused non-importable environment variables (i.e., those
  with names that are not valid shell variable names) to disappear from the
  environment of child processes after invoking more than two external
  commands, potentially crashing the shell. Bug introduced on 2025-05-30.

2026-08-29:

- If an error (such as a syntax error) occurs in a trap action, the error
  message will now have the word "(trap)" in the call stack for clarity.

2026-08-26:

- Following up to 2026-08-16, fixed a remaining problem in the 'read'
  built-in with combined -d and -n options when reading from a pipe; the
  custom delimiter is now correctly used instead of newline.

2026-08-16:

- Fixed a long-standing bug in the 'read' built-in: the -d option was
  ignored when combined with -n. The two options can now be used together.

2026-08-14:

- In the --man/--help output for built-in commands, text marked as italic
  is no longer rendered as bold + underlined, but only as underlined.

2026-08-13:

- Fixed corruption of $_, crashing the shell upon 'typeset -p _', after an
  error occurs trying to redefine a type defined with 'typeset -T'.

2026-08-06:

- Fixed ksh crashing upon parsing a very long list of simple commands
  separated by semicolons or, in 'eval' or dot script contexts, by newlines.

- [v1.1] The -n/--noexec option can now effectively be set in subshells of
  interactive shells, e.g., within a parentheses block.

- [v1.1] The 'eval' command has a new -n/--noexec option that causes it to
  subject the command(s) given to static analysis in the manner of the
  -n/--noexec shell option, emitting warnings for deprecated usages.

2026-08-04:

- Fixed a potential crash when suspending (^Z), on the interactive shell,
  a multi-element pipeline with the last element running in the current
  environment.

- For non-interactive shells, the COLUMNS and LINES variables now initially
  respect their values inherited from the environment, if any. They will
  still be updated upon the next window size change (SIGWINCH) or, on the
  interactive shell, when any external command returns. This change aligns
  ksh with bash in supporting idioms like:
	$ COLUMNS=60 ksh somescript.ksh
	$ COLUMNS=60 print --man

2026-08-03:

- Fixed crashes that could occur when an arithmetic syntax error is thrown
  in an arithmetic expression processed via the print or printf built-ins.

- ksh no longer crashes or misbehaves when an arithmetic expression

src/lib/libast/misc/stk.c
/***********************************************************************
*                                                                      *
*               This software is part of the ast package               *
*          Copyright (c) 1985-2012 AT&T Intellectual Property          *
*          Copyright (c) 2020-2026 Contributors to ksh 93u+m           *
*                      and is licensed under the                       *
*                 Eclipse Public License, Version 2.0                  *
*                                                                      *
*                A copy of the License is available at                 *
*      https://www.eclipse.org/org/documents/epl-2.0/EPL-2.0.html      *
*         (with md5 checksum 84283fa8859daf213bdda5a9f8d1be1d)         *
*                                                                      *
*                 Glenn Fowler <gsf@research.att.com>                  *
*                  David Korn <dgk@research.att.com>                   *
*                   Phong Vo <kpv@research.att.com>                    *
*                  Martijn Dekker <martijn@inlv.org>                   *
*            Johnothan King <johnothanking@protonmail.com>             *
*                                                                      *
***********************************************************************/
/*
 *   Routines to implement a stack-like storage library
 *
 *   A stack consists of a link list of variable size frames
 *   The beginning of each frame is initialized with a frame structure
 *   that contains a pointer to the previous frame and a pointer to the
 *   end of the current frame.
 *
 *   This is a rewrite of the stk library that uses sfio
 *
 *   David Korn
 *   AT&T Research
 *   dgk@research.att.com
 *
 */

#include	<sfio_t.h>
#include	<ast.h>
#include	<align.h>
#include	<stk.h>

/*
 *  A stack is a header and a linked list of frames
 *  The first frame has structure
 *	Sfio_t
 *	Sfdisc_t
 *	struct stk
 * Frames have structure
 *	struct frame
 *	data
 */

#define STK_ALIGN	((size_t)ALIGN_BOUND)
#define STK_FSIZE	(1024*sizeof(char*))
#define STK_HDRSIZE	(sizeof(Sfio_t)+sizeof(Sfdisc_t))

typedef void (*_stk_overflow_)(size_t);
typedef char* (*_old_stk_overflow_)(size_t);	/* for stkinstall (deprecated) */

static int stkexcept(Sfio_t*,int,void*,Sfdisc_t*);
static Sfdisc_t stkdisc = { 0, 0, 0, stkexcept };

Sfio_t	_Stak_data = SFNEW(NULL,0,-1,SFIO_STATIC|SFIO_WRITE|SFIO_STRING,&stkdisc);

struct frame
{
	char	*prev;		/* address of previous frame */
	char	*end;		/* address of end this frame */
	char	**aliases;	/* address aliases */
	ssize_t	nalias;		/* number of aliases */
};

struct stk
{
	_stk_overflow_	stkoverflow;	/* called when malloc fails */
	unsigned int	stkref;		/* reference count */
	int		stkflags;	/* stack attributes */
	char		*stkbase;	/* beginning of current stack frame */
	char		*stkend;	/* end of current stack frame */
};

static size_t		init;		/* 1 when initialized */
static struct stk	*stkcur;	/* pointer to current stk */
static char		*stkgrow(Sfio_t*, size_t);

#define stream2stk(stream)	((stream)==stkstd? stkcur:\
				 ((struct stk*)(((char*)(stream))+STK_HDRSIZE)))
#define stk2stream(sp)		((Sfio_t*)(((char*)(sp))-STK_HDRSIZE))
#define stkleft(stream)		((stream)->_endb-(stream)->_data)

static const char Omsg[] = "out of memory while growing stack\n";

/*
 * default overflow exception
 */
static noreturn void overflow(size_t n)
{
	NoP(n);
	write(2,Omsg, sizeof(Omsg)-1);
	exit(128);
	UNREACHABLE();
}

/*
 * initialize stkstd, sfio operations may have already occurred
 */
static void stkinit(size_t size)
{
	Sfio_t *sp;
	init = size;
	sp = stkopen(0);
	init = 1;
	stkinstall(sp,(_old_stk_overflow_)overflow);
}

static int stkexcept(Sfio_t *stream, int type, void* val, Sfdisc_t* dp)
{
	NoP(dp);
	NoP(val);
	switch(type)
	{
	    case SFIO_CLOSING:
		{
			struct stk *sp = stream2stk(stream);
			if(--sp->stkref == 0)
			{
				if(stream==stkstd)
					stkset(stream,NULL,0);
				else
				{
					void *bp = sp->stkbase, *fp;
					while(fp = bp)
					{
						bp = ((struct frame*)bp)->prev;
						free(fp);
					}
				}
			}
			stream->_data = stream->_next = 0;
		}
		return 0;
	    case SFIO_FINAL:
		free(stream);
		return 1;
	    case SFIO_DPOP:
		return -1;
	    case SFIO_WRITE:
	    case SFIO_SEEK:
		{
			size_t size = (size_t)sfvalue(stream);
			if(init)
			{
				if(!stkgrow(stream, size - (size_t)(stream->_endb - stream->_data)))
					return -1;
			}
			else
				stkinit(size);
		}
		return 1;
	    case SFIO_NEW:
		return -1;
	}
	return 0;
}

/*
 * create a stack
 */
Sfio_t *stkopen(int flags)
{
	size_t bsize;
	Sfio_t *stream;
	struct stk *sp;
	struct frame *fp;
	Sfdisc_t *dp;
	char *cp;
	if(!(stream = calloc(1, sizeof(Sfio_t) + sizeof(Sfdisc_t) + sizeof(struct stk))))
		return NULL;
	dp = (Sfdisc_t*)(stream+1);
	dp->exceptf = stkexcept;
	sp = (struct stk*)(dp+1);
	sp->stkref = 1;
	sp->stkflags = flags;
	sp->stkoverflow = flags&STK_NULL ? 0 : stkcur ? stkcur->stkoverflow : overflow;
	bsize = init+sizeof(struct frame);
	if(flags&STK_SMALL)
		bsize = roundof(bsize,STK_FSIZE/16);
	else
		bsize = roundof(bsize,STK_FSIZE);
	bsize -= sizeof(struct frame);
	if(!(fp = malloc(sizeof(struct frame) + bsize)))
	{
		free(stream);
		return NULL;
	}
	cp = (char*)(fp+1);
	sp->stkbase = (char*)fp;
	fp->prev = 0;
	fp->nalias = 0;
	fp->aliases = 0;
	fp->end = sp->stkend = cp+bsize;
	if(!sfnew(stream,cp,bsize,-1,SFIO_STRING|SFIO_WRITE|SFIO_STATIC|SFIO_EOF))
		return NULL;
	sfdisc(stream,dp);
	return stream;
}

/*
 * return a pointer to the current stack
 * if <stream> is not null, it becomes the new current stack
 * <oflow> becomes the new overflow function
 */
Sfio_t *stkinstall(Sfio_t *stream, _old_stk_overflow_ oflow)
{
	Sfio_t *old;
	struct stk *sp;
	if(!init)
	{
		stkinit(1);
		if(oflow)
			stkcur->stkoverflow = (_stk_overflow_)oflow;
		return NULL;
	}
	old = stkcur?stk2stream(stkcur):0;
	if(stream)
	{
		sp = stream2stk(stream);
		while(sfstack(stkstd, SFIO_POPSTACK));
		if(stream!=stkstd)
			sfstack(stkstd,stream);
		stkcur = sp;
	}
	else
		sp = stkcur;
	if(oflow)
		sp->stkoverflow = (_stk_overflow_)oflow;
	return old;
}

/*
 * set or unset the overflow function
 */
void stkoverflow(Sfio_t *stream, _stk_overflow_ oflow)
{
	struct stk *sp;
	if(!init)
		stkinit(1);
	sp = stream2stk(stream);
	sp->stkoverflow = oflow ? oflow : (sp->stkflags & STK_NULL ? NULL : overflow);
}

/*
 * increase the reference count on the given <stack>
 */
unsigned int stklink(Sfio_t* stream)
{
	struct stk *sp = stream2stk(stream);
	return sp->stkref++;
}

/*
 * terminate a stack and free up the space
 * >0 returned if reference decremented but still > 0
 *  0 returned on last close
 * <0 returned on error
 */
int stkclose(Sfio_t* stream)
{
	struct stk *sp = stream2stk(stream);
	if(sp->stkref>1)
	{
		sp->stkref--;
		return 1;
	}
	return sfclose(stream);
}

/*
 * reset the bottom of the current stack back to <address>
 * if <address> is null, then the stack is reset to the beginning
 * if <address> is not in this stack, the program dumps core
 * otherwise, the top of the stack is set to stkbot+<offset>
 */
void *stkset(Sfio_t *stream, void *address, ptrdiff_t offset)
{
	struct stk *sp = stream2stk(stream);
	char *cp, *loc = (char*)address;
	struct frame *fp;
	unsigned int frames = 0;
	ssize_t n;
	if(!init)
		stkinit((size_t)(offset+1));
	while(1)
	{
		fp = (struct frame*)sp->stkbase;
		cp = sp->stkbase + roundof(sizeof(struct frame), STK_ALIGN);
		n = fp->nalias;
		while(n-->0)
		{
			if(loc==fp->aliases[n])
			{
				loc = cp;
				break;
			}
		}
		/* see whether <loc> is in current stack frame */
		if(loc>=cp && loc<=sp->stkend)
		{
			if(frames)
				sfsetbuf(stream,cp,(size_t)(sp->stkend-cp));
			stream->_data = (unsigned char*)(cp + roundof((size_t)(loc-cp),STK_ALIGN));
			stream->_next = (unsigned char*)loc+offset;
			goto found;
		}
		if(fp->prev)
		{
			sp->stkbase = fp->prev;
			sp->stkend = ((struct frame*)(fp->prev))->end;
			free(fp);
		}
		else
			break;
		frames++;
	}
	/* not found: produce a useful stack trace now instead of a useless one later */
	if(loc)
		abort();
	/* set stack back to the beginning */
	cp = (char*)(fp+1);
	if(frames)
		sfsetbuf(stream,cp,(size_t)(sp->stkend-cp));
	else
		stream->_data = stream->_next = (unsigned char*)cp;
found:
	return stream->_data;
}

/*
 * allocate <n> bytes on the current stack
 *
 * Since 2026-07-29, initialize them to zero.
 */
void *stkalloc(Sfio_t *stream, size_t n)
{
	unsigned char *old;
	if(!init)
		stkinit(n);
	n = roundof(n,STK_ALIGN);
	if(stkleft(stream) <= (ssize_t)n && !stkgrow(stream,n))
		return NULL;
	old = stream->_data;
	stream->_data = stream->_next = old+n;
	return memset(old, 0, n);
}

/*
 * begin a new stack object of at least <n> bytes
 */
void *_stkseek(Sfio_t *stream, ptrdiff_t n)
{
	if(!init)
		stkinit((size_t)n);
	if(stkleft(stream) <= n && !stkgrow(stream,(size_t)n))
		return NULL;
	stream->_next = stream->_data+n;
	return stream->_data;
}

/*
 * advance the stack to the current top
 * if extra is non-zero, first add extra bytes and zero the first
 */
void	*stkfreeze(Sfio_t *stream, size_t extra)
{
	unsigned char *old, *top;
	if(!init)
		stkinit(extra);
	old = stream->_data;
	top = stream->_next;
	if(extra)
	{
		if((ssize_t)extra > (stream->_endb-stream->_next))
		{
			if (!(top = (unsigned char*)stkgrow(stream,extra)))
				return NULL;
			old = stream->_data;
		}
		*top = 0;
		top += extra;
	}
	stream->_next = stream->_data += roundof((size_t)(top-old),STK_ALIGN);
	return (char*)old;
}

/*
 * copy string <str> onto the stack as a new object
 */
char	*stkcopy(Sfio_t *stream, const char* str)
{
	char *cp, *tp;
	size_t n;
	void *base = stkptr(stream,0);
	size_t off = (size_t)stktell(stream);
	/* save unterminated object */
	if(off)
	{
		if (!(tp = malloc(off)))
			return NULL;
		memcpy(tp, stream->_data, off);
	}
	/* allocate space for string on stack */
	n = strlen(str) + 1;
	if (!(cp = stkalloc(stream, n)))
		goto error_out;
	/* restore unterminated object */
	if (off)
	{
		if (!_stkseek(stream,(ptrdiff_t)off))
			goto error_out;
		memcpy(stream->_data, tp, off);
		free(tp);
	}
	/* copy string to stack */
	return memcpy(cp, str, n);

error_out:
	if(off)
		free(tp);
	stkset(stream, base, (ptrdiff_t)off);
	return NULL;
}

/*
 * add a new stack frame of size >= <n> to the current stack.
 * if <n> > 0, copy the bytes from stkbot to stktop to the new stack
 * if <n> is zero, then copy the remainder of the stack frame from stkbot
 * to the end is copied into the new stack frame
 */

static char *stkgrow(Sfio_t *stream, size_t size)
{
	size_t n = size;
	struct stk *sp = stream2stk(stream);
	struct frame *fp= (struct frame*)sp->stkbase;
	char *cp, *dp=0;
	size_t m = (size_t)stktell(stream);
	ptrdiff_t endoff;
	char *end=0, *oldbase=0;
	ssize_t nn=0,add=1;
	n += (m + sizeof(struct frame)+1);
	if(sp->stkflags&STK_SMALL)
		n = roundof(n,STK_FSIZE/16);
	else
		n = roundof(n,STK_FSIZE);
	/* see whether current frame can be extended */
	if(stkptr(stream,0)==sp->stkbase+sizeof(struct frame))
	{
		nn = fp->nalias+1;
		dp=sp->stkbase;
		sp->stkbase = ((struct frame*)dp)->prev;
		end = fp->end;
		oldbase = dp;
		endoff = end - dp;
	}
	cp = realloc(dp, n + (size_t)nn * sizeof(char*));
	if(!cp)
		return NULL;
	if(dp==cp)
	{
		nn--;
		add = 0;
	}
	else if(dp)
	{
		dp = cp;
		end = dp + endoff;
	}
	fp = (struct frame*)cp;
	fp->prev = sp->stkbase;
	sp->stkbase = cp;
	sp->stkend = fp->end = cp+n;
	cp = (char*)(fp+1);
	cp = sp->stkbase + roundof((size_t)(cp-sp->stkbase),STK_ALIGN);
	if(fp->nalias=nn)
	{
		fp->aliases = (char**)fp->end;
		if(end && nn>add)
			memmove(fp->aliases,end,(size_t)(nn-add)*sizeof(char*));
		if(add)
			fp->aliases[nn-1] = oldbase + roundof(sizeof(struct frame),STK_ALIGN);
	}
	if(m && !dp)
		memcpy(cp,(char*)stream->_data,m);
	sfsetbuf(stream,cp,(size_t)(sp->stkend-cp));
	return (char*)(stream->_next = stream->_data+m);
}
